src/activity-logger/background/log-page-visit.js

```diff
@@ -52,17 +56,31 @@ export async function logPageVisit({ tabId, url }) {
     // The time to put in documents.
     const timestamp = Date.now()
 
-    // First create an identifier for the page being visited.
-    const storePageResult = await storePage({ tabId, url })
+    const pageId = generatePageDocId({ url })
+
+    const lookup = initSingleLookup()
+    const dbResult = await lookup(pageId)
+    const latestTime = dbResult == null ? null : dbResult.latest
+
+    let storePageResult
+
+    if (latestTime == null || timestamp - latestTime < threshold) {
+        // First create an identifier for the page being visited.
+        storePageResult = await storePage({ tabId, url })
+    }
 
     // Create a visit pointing to this page (analysing/storing it may still be in progress)
     const { visit } = await storeVisit({
-        page: storePageResult.page,
+        page: { _id: pageId },
         url,
         timestamp,
     })
 
-    await updateIndex(storePageResult.finalPagePromise, visit)
+    if (storePageResult == null) {
+        index.addMetaConcurrent('visit', timestamp, pageId)
```

Comment:
I think best to keep all indexing-related code inside the `updateIndex` function. So you could just pass in `storePageResult` and do the null check inside there and call `index.addMetaConcurrent` if needed